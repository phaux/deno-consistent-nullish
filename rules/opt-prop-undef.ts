import { getRangeWithParens } from "../util/getRangeWithParens.ts";
import { isParenthesized } from "../util/isParenthesized.ts";
import { fc } from "../util/formatCode.ts";

export const optPropUndefRule: Deno.lint.Rule = {
  create(ctx) {
    return {
      TSPropertySignature: visitPropertyNode,
      PropertyDefinition: visitPropertyNode,
    };

    function visitPropertyNode(
      propNode: Deno.lint.TSPropertySignature | Deno.lint.PropertyDefinition,
    ) {
      if (!propNode.optional) return;
      if (!propNode.typeAnnotation) return;
      if (
        propNode.typeAnnotation.typeAnnotation.type !== "TSUnionType" ||
        !propNode.typeAnnotation.typeAnnotation.types.some(
          (t) => t.type === "TSUndefinedKeyword",
        )
      ) {
        const typeAnnotation = propNode.typeAnnotation.typeAnnotation;
        const keyName = ctx.sourceCode.getText(propNode.key);
        return ctx.report({
          node: propNode.key,
          message: `Type of optional property ${fc(keyName)} must include undefined.`,
          hint:
            "TypeScript differentiates missing properties from undefined when `exactOptionalPropertyTypes` option is enabled. " +
            `Include ${fc("undefined")} in the type to represent optional properties.`,
          *fix(fixer) {
            if (
              isParenthesized(ctx.sourceCode, typeAnnotation) ||
              typeAnnotation.type.endsWith("Keyword") ||
              typeAnnotation.type === "TSLiteralType" ||
              typeAnnotation.type === "TSTemplateLiteralType" ||
              typeAnnotation.type === "TSTypeReference" ||
              typeAnnotation.type === "TSIndexedAccessType" ||
              typeAnnotation.type === "TSArrayType" ||
              typeAnnotation.type === "TSTypeLiteral" ||
              typeAnnotation.type === "TSTupleType" ||
              typeAnnotation.type === "TSUnionType"
            ) {
              // Type annotation is a simple type.
              yield fixer.insertTextAfterRange(
                getRangeWithParens(ctx.sourceCode, typeAnnotation),
                " | undefined",
              );
            } else {
              // Type annotation is a complex type.
              // Wrap the complex type in parentheses to make sure precedence is correct.
              yield fixer.insertTextBefore(typeAnnotation, "(");
              yield fixer.insertTextAfter(typeAnnotation, ") | undefined");
            }
          },
        });
      }
    }
  },
};
