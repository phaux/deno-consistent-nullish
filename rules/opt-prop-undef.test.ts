import { assertSnapshot } from "@std/testing/snapshot";
import { assertEquals } from "@std/assert";
import { optPropUndefRule } from "./opt-prop-undef.ts";

const plugin: Deno.lint.Plugin = {
  name: "test",
  rules: { rule: optPropUndefRule },
};

Deno.test("requires optional properties to include undefined", async (t) => {
  const diagnostics = Deno.lint.runPlugin(
    plugin,
    "test.ts",
    `
    type T1 = { a?: string }
    interface T2 { a?: string | number }
    class C { a?: () => undefined }
    `,
  );
  assertEquals(diagnostics.length, 3);
  await assertSnapshot(t, diagnostics);
});

Deno.test("allows optional properties which include undefined", () => {
  const diagnostics = Deno.lint.runPlugin(
    plugin,
    "test.ts",
    `
    type T1 = { a?: string | undefined }
    interface T2 { a?: string | number | undefined }
    class C { a?: (() => undefined) | undefined }
    `,
  );
  assertEquals(diagnostics.length, 0);
});
