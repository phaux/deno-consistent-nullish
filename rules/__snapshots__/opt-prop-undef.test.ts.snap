export const snapshot = {};

snapshot[`requires optional properties to include undefined 1`] = `
[
  {
    fix: [
      {
        range: [
          27,
          27,
        ],
        text: " | undefined",
      },
    ],
    hint: "TypeScript differentiates missing properties from undefined when \`exactOptionalPropertyTypes\` option is enabled. Include \`undefined\` in the type to represent optional properties.",
    id: "test/rule",
    message: "Type of optional property \`a\` must include undefined.",
    range: [
      17,
      18,
    ],
  },
  {
    fix: [
      {
        range: [
          68,
          68,
        ],
        text: " | undefined",
      },
    ],
    hint: "TypeScript differentiates missing properties from undefined when \`exactOptionalPropertyTypes\` option is enabled. Include \`undefined\` in the type to represent optional properties.",
    id: "test/rule",
    message: "Type of optional property \`a\` must include undefined.",
    range: [
      49,
      50,
    ],
  },
  {
    fix: [
      {
        range: [
          89,
          89,
        ],
        text: "(",
      },
      {
        range: [
          104,
          104,
        ],
        text: ") | undefined",
      },
    ],
    hint: "TypeScript differentiates missing properties from undefined when \`exactOptionalPropertyTypes\` option is enabled. Include \`undefined\` in the type to represent optional properties.",
    id: "test/rule",
    message: "Type of optional property \`a\` must include undefined.",
    range: [
      85,
      86,
    ],
  },
]
`;
